### PR TITLE
Nomic Client 3.0.0

### PR DESCRIPTION
**Pull request by AndriyMulyar**
Nomic Client 3.0.0

### New Features
- All identifiers moved to unique, human-readable, URL valid organization/project slugs. These are auto-created for you on dataset creation and are ensured to be unique across your Atlas organization
```python
from nomic import AtlasDataset
dataset = AtlasDataset('my-organization/my-dataset')
print(dataset.maps[0])

```
- AtlasProject renamed to AtlasDataset
- Makes supplying an index_name optional (by being set to None)
- map_data unifies easy interfaces for indexing datasets.

### Deprecations:
- map_embeddings, map_text in favor a single map_data
- Deprecates iterable in map_text, map_data will no longer support this iterable workflow. Use AtlasDataset and .add_data instead.

### Transitioning from 2.x to 3.x
- Rename all map_text and map_embedding calls to map_data
- Replace any use of AtlasProject with AtlasDataset
- See examples in the python client examples folder for details.

### CODE REVIEW

File: nomic/atlas.py [hunk 1]
```diff
@@ -28,35 +28,33 @@ def map_embeddings(
...
-        name: A name for your map.
+        name: A name for your dataset. Specify in the format `organization/project` to create in a specific organization.
         description: A description for your map.
         is_public: Should this embedding map be public? Private maps can only be accessed by members of your organization.
-        colorable_fields: The project fields you want to be able to color by on the map. Must be a subset of the projects fields.
-        organization_name: The name of the organization to create this project under. You must be a member of the organization with appropriate permissions. If not specified, defaults to your user accounts default organization.
-        reset_project_if_exists: If the specified project exists in your organization, reset it by deleting all of its data. This means your uploaded data will not be contextualized with existing data.
-        add_datums_if_exists: If specifying an existing project and you want to add data to it, set this to true.
+        colorable_fields: The dataset fields you want to be able to color by on the map. Must be a subset of the projects fields.
```
Comment (bmschmidt) [inline nomic/atlas.py:47]: For pushing out a 3.0 release, we could simply deprecateWarning colorable_fields, with a plan to remove… and then going forward it makes more sense to just deprecateWarning this when passed rather than remove to avoid breaking old code.

File: nomic/atlas.py [hunk 2]
```diff
@@ -28,35 +28,33 @@ def map_embeddings(
     build_topic_model: bool = True,
     topic_label_field: str = None,
     num_workers: None = None,
-    organization_name: str = None,
```
[the remark this answers is not in the document]
Comment (AndriyMulyar) [reply nomic/atlas.py:31]: done

File: nomic/atlas.py (hunk 1 above)
Comment (AndriyMulyar) [reply nomic/atlas.py:182]: sounds good

File: nomic/atlas.py [hunk 3]
```diff
@@ -12,10 +12,144 @@
...
-from .project import AtlasProject
+from .dataset import AtlasDataset
 from .settings import *
 from .utils import b64int, get_random_name, arrow_iterator
 
+from .data_inference import NomicProjectHyperparameters, NomicTopicHyperparameters, NomicDuplicatesHyperparameters
+
+
+def map_data(
+    data: List[Dict] = None,
```
Comment (bmschmidt) [inline nomic/atlas.py:23]: map_data should consume pandas df as well as list[Dict]. It looks cooler in examples, and pa.Table.from_pandas(df) tends to be better in every way (speed, memory, type safety) than pa.Table.from_pylist(lod).

Might as well consume arrow tables too, but not worth bragging about.

File: nomic/atlas.py [hunk 4]
```diff
@@ -12,10 +12,144 @@
...
+    embeddings: np.array = None,
+    name: str = None,
+    description: str = "",
+    id_field: str = None,
+    is_public: bool = True,
+    indexed_field: str = None,
+    projection: Optional[Dict] = None,
+    topic_model: Optional[Dict] = None,
+    duplicate_detection: Optional[Dict] = None
```
Comment (bmschmidt) [inline nomic/atlas.py:32]: Rather than `None`, what if this was `topic_model: Union[True, False, TopicModelOptions]`. True means build a topic model with default settings: False means don't build a topic model: passing a dict turns it into `**kwargs`.

I never love having `None` mean "do something", and it also keeps the ability to turn off modeling.

File: nomic/atlas.py [hunk 5]
```diff
@@ -12,10 +12,144 @@
...
+
+    if indexed_field is not None:
+        modality = 'text'
+
+    if id_field is None:
+        id_field = ATLAS_DEFAULT_ID_FIELD
+
+    project_name = get_random_name()
+    if description is None:
+        description = 'A description for your map.'
```
Comment (bmschmidt) [inline nomic/atlas.py:63]: Description cannot be `None` according the type signature. I think we should remove this line anyway, an empty string or NULL or better than this placeholder.

File: nomic/atlas.py [hunk 6]
```diff
@@ -12,10 +12,144 @@
...
+        data = [{ATLAS_DEFAULT_ID_FIELD: b64int(i)} for i in range(len(embeddings))]
+        added_id_field = True
+
+
+    if id_field == ATLAS_DEFAULT_ID_FIELD and id_field not in data[0]:
+        added_id_field = True
+        for i in range(len(data)):
+            # do not modify object the user passed in - also ensures IDs are unique if two input datums are the same *object*
+            data[i] = data[i].copy()
+            data[i][id_field] = b64int(i)
```
[the remark this answers is not in the document]
Comment (bmschmidt) [reply nomic/atlas.py:84]: Actually--things would be cleaner if we moved auto-id generation right into the add_data method. Then the data kwarg could be passed straight through, keeping the union type on add_data; maybe the default for project.add_data would be to raise on not having an id_field and we flip that to generate with an argument, or maybe we let auto-id-generation into those functions.

File: nomic/atlas.py [hunk 7]
```diff
@@ -12,10 +12,144 @@
...
+        duplicate_detection=duplicate_detection.tag_duplicates,
+        duplicate_threshold=duplicate_detection.duplicate_cutoff,
+    )
+
+    project = dataset._latest_project_state()
+    return project
+
+
 
 def map_embeddings(
```
[the remark this answers is not in the document]
Comment (bmschmidt) [reply nomic/atlas.py:154]: Oh never mind, deprecated now I remember.

File: nomic/atlas.py [hunk 8]
```diff
@@ -216,16 +364,15 @@ def map_text(
...
+        logger.warning("Passed data has no samples. No dataset will be created")
         return
 
-    project = AtlasProject(
-        name=project_name,
+    project = AtlasDataset(
+        identifier=project_name,
         description=description,
         unique_id_field=id_field,
         modality='text',
```
Comment (bmschmidt) [inline nomic/atlas.py:374]: Can we make the necessary atlas_cloud changes to drop modality from the AtlasProject arguments? Even if we retain projects as having necessary modalities for the time being, we can defer deciding what the type of a project is until the user calls `upload_text` or `upload_embeddings`. I think the plan I wrote [here](https://www.notion.so/nomic-ai/Multiple-Modalities-in-Atlas-9cb602a28e2d47d9b67a6adf12379035) is still workable.

File: nomic/dataset.py [hunk 9]
```diff
@@ -156,10 +161,74 @@ def _get_project_by_id(self, project_id: str):
...
+        return response.json()
+
+    def _get_organization_by_slug(self, slug: str):
+        '''
+
+        Args:
+            slug: The organization slug
+
+        Returns:
+            Returns the requested project.
```
Comment (bmschmidt) [inline nomic/dataset.py:175]: Would be better to merge upstream and just hit f"/v1/organization/{slug}" for full organization info directly, avoiding a roundtrip.
```suggestion
            An organization id
```

File: nomic/dataset.py [hunk 10]
```diff
@@ -156,10 +161,74 @@ def _get_project_by_id(self, project_id: str):
...
+
+        organization_slug = identifier.split('/')[0]
+        project_slug = identifier.split('/')[1]
+        response = requests.get(
+            self.atlas_api_path + f"/v1/project/{organization_slug}/project/{project_slug}",
+            headers=self.header,
+        )
+
+        if response.status_code != 200:
+            return None
```
Comment (bmschmidt) [inline nomic/dataset.py:212]: Shouldn't this raise an error?

File: nomic/dataset.py [hunk 11]
```diff
@@ -156,10 +161,74 @@ def _get_project_by_id(self, project_id: str):
...
+
+        if not self.is_valid_dataset_identifier(identifier=identifier):
+            raise Exception("Invalid dataset identifier")
+
+        organization_slug = identifier.split('/')[0]
+        project_slug = identifier.split('/')[1]
+        response = requests.get(
+            self.atlas_api_path + f"/v1/project/{organization_slug}/project/{project_slug}",
+            headers=self.header,
+        )
```
Comment (bmschmidt) [inline nomic/dataset.py:209]: I want to get rid of that `/project/` route in favor of the shorter format.
```suggestion
        response = requests.get(
            self.atlas_api_path + f"/v1/project/{identifier}",
            headers=self.header,
        )
```

File: nomic/dataset.py [hunk 12]
```diff
@@ -156,10 +161,74 @@ def _get_project_by_id(self, project_id: str):
...
 
         return response.json()
 
+
+    def is_valid_dataset_identifier(self, identifier: str):
+        '''
+        Checks if a string is a valid identifier for a dataset
+
+        Args:
+            identifer: the organization slug and dataset slug seperated by a slash
```
Comment (bmschmidt) [inline nomic/dataset.py:222]:
```suggestion
            identifer: the organization slug and dataset slug separated by a slash
```

File: nomic/dataset.py (hunk 11 above)
Comment (AndriyMulyar) [reply nomic/dataset.py:209]: yeah that didn't exist when i wrote this code :)

File: nomic/dataset.py [hunk 13]
```diff
@@ -656,87 +687,86 @@ def _get_atoms(self, ids: List[str]) -> List[Dict]:
...
-        Atlas projects store data (text, embeddings, etc) that you can organize by building indices.
-        If the organization already contains a project with this name, it will be returned instead.
+        Creates or loads an AtlasDataset.
+        AtlasDataset's store data (text, embeddings, etc) that you can organize by building indices.
+        If the organization already contains a dataset with this name, it will be returned instead.
 
         **Parameters:**
 
-        * **project_name** - The name of the project.
+        * **identifier** - The dataset identifier (format is `organization/project`) or just dataset and we will auto-infer from the default org.
```
Comment (bmschmidt) [inline nomic/dataset.py:710]:
```suggestion
        * **identifier** - The dataset identifier in the form `project` or `organization/project`. If no organization is passed, your default organization will be used.
```

File: nomic/atlas.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AndriyMulyar) [reply nomic/atlas.py:84]: Theres an issue with add_data here in that it works with batching. I would prefer to keep this logic in the quickmap method.

File: nomic/dataset.py [hunk 14]
```diff
@@ -656,87 +687,86 @@ def _get_atoms(self, ids: List[str]) -> List[Dict]:
...
         * **description** - A description for the project.
```
Comment (bmschmidt) [inline nomic/dataset.py:711]: It's weird to allow passing the *description* here but not the project name that appears on the top. We could allow passing the display name as an argument, but by default use the slug and not allow fetching by display name.

File: nomic/dataset.py [hunk 15]
```diff
@@ -656,87 +687,86 @@ def _get_atoms(self, ids: List[str]) -> List[Dict]:
...
 
-        results = self._get_existing_project_by_name(project_name=name, organization_name=organization_name)
+        project = self._get_project_by_slug_identifier(identifier=identifier)
 
-        if 'project_id' in results:  # project already exists
-            organization_name = results['organization_name']
-            project_id = results['project_id']
+        if project:  # dataset already exists
+            project_id = project['id']
             if reset_project_if_exists:  # reset the project
```
Comment (bmschmidt) [inline nomic/dataset.py:740]: FWIW I still dislike this flag and think we should get rid of it.

File: nomic/dataset.py [hunk 16]
```diff
@@ -656,87 +687,86 @@ def _get_atoms(self, ids: List[str]) -> List[Dict]:
...
+
         if project_id is not None:
             self.meta = self._get_project_by_id(project_id)
             return
 
-        if organization_name is None:
-            organization_name = self._get_current_users_main_organization()['nickname']
+        if not self.is_valid_dataset_identifier(identifier=identifier):
+            default_org_slug = self._get_current_users_main_organization()['slug']
+            identifier = default_org_slug+'/'+identifier
```
Comment (bmschmidt) [inline nomic/dataset.py:734]: Do we have validation that the identifier passes `re.match("^[a-z\-0-9]+$")` anywhere?

File: nomic/tests/test_atlas_client.py [hunk 17]
```diff
@@ -271,7 +260,7 @@ def test_data():
 
 
 def test_interactive_workflow():
-    p = AtlasProject(name='UNITTEST1', modality='text', unique_id_field='id', reset_project_if_exists=True)
+    p = AtlasDataset('UNITTEST1', modality='text', unique_id_field='id')
```
Comment (bmschmidt) [inline nomic/tests/test_atlas_client.py:263]: Not a valid slug

File: nomic/atlas.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bmschmidt) [reply nomic/atlas.py:84]: Oh right forgot about that. Makes sense then.

File: nomic/dataset.py [hunk 18]
```diff
@@ -156,10 +161,74 @@ def _get_project_by_id(self, project_id: str):
...
+        Args:
+            identifer: the organization slug and dataset slug seperated by a slash
+
+        Returns:
+            Returns the requested project.
+        '''
+        slugs = identifier.split('/')
+        if '/' not in identifier or len(slugs) != 2:
+            return False
+        return True
```
[the remark this answers is not in the document]
Comment (bmschmidt) [reply nomic/dataset.py:230]: On second thought it should be a method: but it should also validate that the parts pass the slug regex "^[a-z\-0-9]+$"

File: nomic/dataset.py (hunk 15 above)
Comment (AndriyMulyar) [reply nomic/dataset.py:740]: yep

**Comment by bmschmidt** [comment]
I just took a while understanding why my code was broken. We should add something like

```
class AtlasProject():
    def __init__(**kwargs):
        raise DeprecationWarning("Atlas Project has been replaced by AtlasDataset")
        super.init(**kwargs)
```

File: nomic/atlas.py (hunk 8 above)
Comment (AndriyMulyar) [reply nomic/atlas.py]: done

File: nomic/atlas.py (hunk 4 above)
Comment (AndriyMulyar) [reply nomic/atlas.py]: Done

File: nomic/atlas.py (hunk 3 above)
Comment (AndriyMulyar) [reply nomic/atlas.py]: done

File: examples/map_text.py [hunk 19]
```diff
@@ -8,10 +8,9 @@
 subset_idxs = np.random.choice(len(dataset), size=max_documents, replace=False).tolist()
 documents = [dataset[i] for i in subset_idxs]
 
-project = atlas.map_text(data=documents,
+project = atlas.map_data(data=documents,
                           indexed_field='text',
                           name='News 10k Example',
-                          colorable_fields=['label'],
```
Comment (bmschmidt) [inline examples/map_text.py:14]: Why is this removed?

File: examples/map_text_and_get_topics.py [hunk 20]
```diff
@@ -10,18 +10,17 @@
 subset_idxs = np.random.choice(len(dataset), size=max_documents, replace=False).tolist()
 documents = [dataset[i] for i in subset_idxs]
 
-project = atlas.map_text(data=documents,
+project = atlas.map_data(data=documents,
                           indexed_field='text',
                           name='News 10k For Topic Extraction',
-                          colorable_fields=['label'],
```
Comment (bmschmidt) [inline examples/map_text_and_get_topics.py:16]: ditto

File: examples/map_text_with_nomic_embed.py [hunk 21]
```diff
@@ -41,5 +41,5 @@ def generate_embeddings(documents):
 document_embeddings = generate_embeddings(documents)
 
 print(document_embeddings.shape)
-response = atlas.map_embeddings(embeddings=document_embeddings, data=documents, topic_label_field='text', build_topic_model=True)
+response = atlas.map_data(embeddings=document_embeddings, data=documents, topic_model=dict(topic_label_field='text'))
```
Comment (bmschmidt) [inline examples/map_text_with_nomic_embed.py:44]:
```suggestion
response = atlas.map_data(embeddings=document_embeddings, data=documents, topic_model={"topic_label_field" : "text"}))
```

Comment (bmschmidt) [reply examples/map_text_with_nomic_embed.py:44]: Putting `topic_label_field` as an argument inside `topic_model` feels redundant--should possibly be just `label_field`.

File: nomic/atlas.py [hunk 22]
```diff
@@ -12,10 +12,126 @@
...
+        description = description
+
+    # no metadata was specified
+    added_id_field = False
+    if data is None:
+        data = [{ATLAS_DEFAULT_ID_FIELD: b64int(i)} for i in range(len(embeddings))]
+        added_id_field = True
+
+
+    if id_field == ATLAS_DEFAULT_ID_FIELD and id_field not in data[0]:
```
Comment (bmschmidt) [inline nomic/atlas.py:78]: I think this line will break if `data` is a pandas df.

File: nomic/atlas.py [hunk 23]
```diff
@@ -28,40 +144,47 @@ def map_embeddings(
...
 
     '''
 
     assert isinstance(embeddings, np.ndarray), 'You must pass in a numpy array'
 
+    raise DeprecationWarning("map_embeddings is deprecated and will soon be removed, use atlas.map_data instead.")
+
+    if organization_name is not None:
+        logger.warning(
+            "Passing organization name has been removed in Nomic Python client 3.0. Instead identify your dataset with `organization_name/project_name` (e.g. sterling-cooper/november-ads).")
```
Comment (bmschmidt) [inline nomic/atlas.py:184]:
```suggestion
            "Passing organization name has been removed in Nomic Python client 3.0. Instead identify your dataset with `organization_name/project_name` (e.g. "sterling-cooper/november-ads").")
```

File: nomic/dataset.py (hunk 10 above)
Comment (bmschmidt) [reply nomic/dataset.py:211]: I'm still confused why this returns None instead of throwing.

File: nomic/utils.py [hunk 24]
```diff
@@ -1,10 +1,33 @@
...
+    'hypatia', 'leibniz', 'copernicus', 'hubble', 'darwin', 'hooke', 'faraday', 'pasteur', 'mendel', 'galvani',
+    'bell', 'edison', 'tesla', 'boltzmann', 'dirac', 'fermi', 'oppenheimer', 'schrodinger', 'fibonacci',
+    'fermat', 'cantor', 'jordan', 'riemann', 'babbage', 'dijkstra', 'shannon', 'erdos', 'boole',
+    'lagrange', 'lebesgue', 'legendre', 'gauss', 'cauchy', 'ramanujan', 'weierstrass', 'fourier',
+    'cavendish', 'lavoisier', 'becquerel', 'avogadro', 'dalton', 'rutherford', 'thomson', 'bose', 'fisher',
+    'jaynes', 'bernoulli', 'alembert', 'gibbs', 'pareto', 'levy', 'sierpinski', 'cohen', 'hardy', 'wigner',
+    'mercator', 'ortelius', 'ptolemy', 'speed', 'schedel', 'waldseemuller', 'homann', 'rennell', 'ogilby', 'hondius',
+    'moll', 'blaeu', 'sanson', 'borgonio', 'fer', 'isle', 'munster', 'hollar', 'kitchin',
+    'lecun', 'bengio', 'hinton', 'ng', 'schmidhuber', 'fei-fei', 'koller', 'russell', 'norvig', 'vapnik',
+    'bishop', 'goodfellow', 'kingma', 'arora', 'sutton', 'breiman'
```
Comment (bmschmidt) [inline nomic/utils.py:21]: This list is 95.8% male.

Comment (bmschmidt) [inline nomic/utils.py:21]:
```suggestion
    'newton', 'einstein', 'gauss', 'pascal', 'laplace',
    'euler', 'kepler', 'wiles', 'neumann', 'noether', 'bohr', 'heisenberg', 'planck', 'maxwell', 'hawking',
    'feynman', 'curie', 'poincare', 'turing', 'nash', 'hilbert', 'godel', 'galois', 'ramanujan', 'lovelace',
    'hypatia', 'leibniz', 'copernicus', 'hubble', 'darwin', 'hooke', 'faraday', 'pasteur', 'mendel', 'galvani',
    'bell', 'boltzmann', 'dirac', 'fermi', 'oppenheimer', 'schrodinger', 'fibonacci',
    'chatelet', 'herschel', 'kovalevskaya', 'somerville', 'franklin', 'lepore', 'yalow',
     'jemison', 'burnell', 'bartik', 'spencer', 'hamilton', 'lovelace', 'johnson', 'ride', 'anning', 'meitner',
     'hopper', 'mcclintock', 'franklin', 'noether', 'rubin', 'wu',
    'fermat', 'cantor', 'jordan', 'riemann', 'babbage', 'dijkstra', 'shannon', 'erdos', 'boole',
    'lagrange', 'lebesgue', 'legendre', 'gauss', 'cauchy', 'ramanujan', 'weierstrass', 'fourier',
    'cavendish', 'lavoisier', 'becquerel', 'avogadro', 'dalton', 'rutherford', 'thomson', 'bose', 'fisher',
    'jaynes', 'bernoulli', 'alembert', 'gibbs', 'pareto', 'levy', 'sierpinski', 'hardy', 'wigner',
    'mercator', 'ortelius', 'speed', 'schedel', 'waldseemuller', 'homann', 'rennell', 'ogilby', 'hondius',
    'moll', 'blaeu', 'sanson', 'borgonio', 'isle', 'munster', 'hollar', 'kitchin',
    'lecun', 'bengio', 'hinton', 'schmidhuber', 'fei-fei', 'koller', 'russell', 'norvig', 'vapnik',
    'bishop', 'goodfellow', 'kingma', 'arora', 'sutton', 'breiman'
```

File: nomic/utils.py [hunk 25]
```diff
@@ -1,10 +1,33 @@
...
+]
+
+adjectives = [
+    'curious', 'innovative', 'analytical', 'creative', 'logical', 'adventurous', 'inquisitive', 'imaginative',
+    'dynamic', 'quirky', 'resourceful', 'persistent', 'observant', 'insightful', 'intuitive', 'experimental',
+    'inspiring', 'inventive', 'tenacious', 'stubborn', 'lazy', 'disorganized', 'inattentive', 'careless',
+    'unpredictable', 'irrational', 'impulsive', 'chaotic', 'indecisive', 'cynical', 'neglectful', 'overbearing',
+    'oblivious', 'inflexible'
```
Comment (bmschmidt) [inline nomic/utils.py:29]:
```suggestion
    'oblivious', 'inflexible', 'lackadaisical', 'truculent', 'diffident', 'lumbering'
```

File: setup.py [hunk 26]
```diff
@@ -26,11 +26,9 @@
         'numpy',
         'pandas',
         'pydantic',
-        'wonderwords',
         'tqdm',
-        'cohere',
         'pyarrow',
-	'pillow',
+        'pillow',
```
Comment (bmschmidt) [inline setup.py:31]: I wish we didn't need pillow

File: nomic/tests/test_atlas_client.py [hunk 27]
```diff
@@ -400,23 +364,20 @@ def test_map_text_pandas():
         'color': [random.choice(['red', 'blue', 'green']) for i in range(size)],
     })
 
-    project = atlas.map_text(
+    dataset = atlas.map_data(
         name='UNITTEST_pandas_text',
         id_field='id',
```
Comment (bmschmidt) [inline nomic/tests/test_atlas_client.py:369]: We should test the ability to auto-generate ids in the pandas test--this has bit us before.

```suggestion
```

**Review by bmschmidt — COMMENTED** [review]
I suspect that map_data with pandas is broken rn.

**Comment by apage43** [comment]
tried to run some old code w/ this branch and hit 

```
DeprecationWarning: map_embeddings is deprecated and will soon be removed, use atlas.map_data instead.
```

when I tried to use `map_text` - if `map_text` is also being deprecated we should fix the msg here

**Comment by rguo123** [comment]
Not blocking but there are a lot of references to `project` throughout the nomic client. If we're phasing out AltasProject to AtlasDataset, it'll be good code hygiene to reference these as `dataset` as well. Started this in DownloadTags PR

**Comment by AndriyMulyar** [comment]
> Not blocking but there are a lot of references to `project` throughout the nomic client. If we're phasing out AltasProject to AtlasDataset, it may be good code hygiene to change these to `dataset` as well at least in the python client. Started this in DownloadTags PR.

did this heavily on documentation, internal facing objects i've left as is because the api URI's still use project

**Review by bmschmidt — APPROVED** [review]
